### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by mariano-aguero**
Added some improvements:
- Improve readme, add pre-requisites, versions, setting up dev.
- Add circleci
- Add MIT License
- Add error messages for Token specification based on [FA1.2 standard ](https://gitlab.com/tzip/tzip/-/tree/master/proposals/tzip-7)
- Prevent the attack scenario as described in FA1.2 standard corresponding [attack vector](https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/edit)

For more, see comments

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,26 +1,41 @@
-### tezos-defi-dapp
+# Tezos defi dapp
 User interface with high level of deFi functionalities 
 
+## Getting started
 
-### Download faucet ...
+A quick introduction of the minimal setup you need to get Tezos defi dapp up & running.
+
+
```
Comment (mariano-aguero) [inline README.md:8]: Improve readme with prerequisites and versions

File: contracts/fa12.ligo [hunk 2]
```diff
@@ -78,26 +78,36 @@ function approve (const addressSpender : address; const value : nat; var store :
     if sender = addressSpender then skip;
     else block {
         const senderAccount: account = getAccount(sender, store.accounts);
-        senderAccount.allowances[addressSpender] := value;
-        store.accounts[sender] := senderAccount;
+        var allowed: nat :=  getAllowance(addressSpender, senderAccount.allowances); 
+
+        // Changing allowance value from non-zero value to a non-zero value is forbidden to prevent the corresponding attack vector.
+        if allowed =/= 0n then failwith("UnsafeAllowanceChange");
```
Comment (mariano-aguero) [inline contracts/fa12.ligo:84]: Check vector attack right here, added a proper error message based on tzip

File: contracts/fa12.ligo [hunk 3]
```diff
@@ -78,26 +78,36 @@ function approve (const addressSpender : address; const value : nat; var store :
...
+    // the user transfers their own money and this does not require approval.
     if addressFrom = addressTo then skip;
     else block {
-      // Check if accountFrom allowed to spend value
+      // #2 NotEnoughAllowance: the transaction sender must be previously authorized to transfer at 
+      // least the requested number of tokens from the "from" account using the approve
       case isAllowed(addressFrom, addressTo, value, store) of 
-      | False -> failwith ("Sender not allowed to spend token")
+      | False -> block {
+          failwith ("NotEnoughAllowance");
```
Comment (mariano-aguero) [inline contracts/fa12.ligo:102]: New error message for not enough allowance

File: contracts/fa12.ligo [hunk 4]
```diff
@@ -78,26 +78,36 @@ function approve (const addressSpender : address; const value : nat; var store :
...
+        }
       | True -> skip
       end;
 
       const addressFromAccount: account = getAccount(addressFrom, store.accounts);
-      // Check that the accountFrom can spend that much
+      // #3 NotEnoughBalance: check that the accountFrom can spend that much
       if value > addressFromAccount.balance
-      then failwith ("Balance is too low");
+      then failwith ("NotEnoughBalance");
```
Comment (mariano-aguero) [inline contracts/fa12.ligo:110]: New error message for not enough balance

File: contracts/fa12.ligo [hunk 5]
```diff
@@ -182,7 +192,7 @@ function burn (const value : nat ; var store : store) : return is
 
     // Check that the owner can spend that much
     if value > ownerAccount.balance 
-    then failwith ("Owner balance is too low");
+    then failwith ("NotEnoughBalance");
```
Comment (mariano-aguero) [inline contracts/fa12.ligo:195]: Idem right here for error message updates

File: contracts/fa12.ligo [hunk 6]
```diff
@@ -198,12 +208,43 @@ function burn (const value : nat ; var store : store) : return is
  } with (emptyOps, store)
 
 
+function burnTo (const toAddress: address ; const value : nat ; var store : store) : return is
```
Comment (mariano-aguero) [inline contracts/fa12.ligo:211]: This function is used in the withdraw operation in the pool.ligo function

File: contracts/pool.ligo [hunk 7]
```diff
@@ -2,31 +2,40 @@
 #include "./utils/math.ligo"
 #include "./partials/tokenActions.ligo"
 
-type action is
```
Comment (mariano-aguero) [inline contracts/pool.ligo:5]: Type moved to a shared file

File: package.json [hunk 8]
```diff
@@ -2,6 +2,10 @@
   "name": "tezos-defi-dapp",
   "version": "0.1.0",
   "private": true,
+  "engines" : {
```
Comment (mariano-aguero) [inline package.json:5]: Added proper versions for engines

**Comment by cmalfesi** [comment]
The changes look good.

**Pull request merged**